lib/pharos/phase_manager.rb

```diff
@@ -19,25 +25,37 @@ def initialize(ssh_manager:, **options)
 
     # @param phases [Array<Pharos::Phases::Base>]
     # @return [Array<...>]
-    def run_parallel(phases)
+    def run_parallel(phases, &block)
       threads = phases.map { |phase|
         Thread.new do
-          begin
-            yield phase
-          rescue StandardError => exc
-            puts " [#{phase}] #{exc.class}: #{exc.message}"
-            raise
-          end
+          yield_phase_with_retry(phase, &block)
         end
       }
       threads.map(&:value)
     end
 
     # @param phases [Array<Pharos::Phases::Base>]
     # @return [Array<...>]
-    def run_serial(phases)
+    def run_serial(phases, &block)
       phases.map do |phase|
+        yield_phase_with_retry(phase, &block)
+      end
+    end
+
+    # @param phase [Pharos::Phases::Base]
+    # @param retry_times [Integer]
+    def yield_phase_with_retry(phase, retry_times = 10)
+      retries = 0
+      begin
         yield phase
+      rescue *RETRY_ERRORS => exc
+        raise if retries >= retry_times
+
+        logger.error { "[#{phase.host}] got error (#{exc.class.name}) #{exc.message}" }
```

Comment:
`logger.debug{ "#{exc.backtrace.join("\n")}" }` would be a good idea for debugging errors during development.

Reply:
Debug backtrace is now there.